ECIP-1055: Succinct PoW Using Merkle Mountain Ranges (#11)
etclabscore/ECIPs: ECLIPs/ECIP-draft_succinct-proofs.md

---

> eip: <to be assigned>

Reviewer: ```suggestion
ecip: 1055
```

why did you use the EIP template not the ECIP template?

Author: my thought was that ETH might want to do it too, and that I would submit it as both EIP and ECIP

- A list of the randomly selected ~500 blockheaders for the proof
  - a standard deterministic way to pick the 500 hashes (using blockhash as seed) over the probability denisty function. I propose that we do something like [this](https://en.wikipedia.org/wiki/Slice_sampling)
  - The 64 byte dataset (each) for verifying an ethHash (~2.4 MB)
- A list of proofNodes from the MMR (+"bag proof" nodes)

## Rationale
<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->

## Backwards Compatibility
<!--All EIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The EIP must explain how the author proposes to deal with these incompatibilities. EIP submissions without a sufficient backwards compatibility treatise may be rejected outright.-->

> All EIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The EIP must explain how the author proposes to deal with these incompatibilities. EIP submissions without a sufficient backwards compatibility treatise may be rejected outright.

Reviewer: WIP?

Author: yeah thats a mistake

<!-- at all unless you check its proof of work. Doing that would yeild a cost parameter:

- `block_x` is proven to have cost about `5 Ether` to produce.

nowing that something cost 5 Ether is not sufficient for most applications. By checking a suffix (blocks after this point in the chain) we can prove that the data cost at least `5n Ether` to produce, where n is the number of blocks checked.

With the simple assumption that the honest chain has 2/3 of the total mining power

The next thing we can do is to request the header before that and validate its PoW as well -->

The problem is that it ends there. And we still have nothing, if the person (or machine) validating cannot validate the block. The cost to produce such proofs is free; To validate blocks we cant use the same strategy. Blocks validate against a chain, which in turn validates through its proof of work. Currently the only way to do this requires 4gb of chain data. Thats way too much. We need something that can work within a few seconds. 

FlyClient is that something, and adding an `MMRroot` to each blockheader is the protocol change that will allow FlyClient to work.

> On a high level, it allows the verifier to sample a subset of blocks in a tree, rather than a full amount thought a linked list, improving the time-complexity from O(n) to O(log(n)). The sample subset can be made large enough such it is cryptographically infeasible to ever produce a false proof.

Reviewer 2: ```suggestion
On a high level, it allows the verifier to sample a subset of blocks in a tree, rather than a full amount through a linked list, improving the time-complexity from O(n) to O(log(n)). The sample subset can be made large enough such it is cryptographically infeasible to ever produce a false proof.
```

There is a consensus-breaking rule change proposed here, and then a bit of software around the change will be needed in order to take advantage of the new features. I think it would be best to keep _this_ EIP as just detailing the rule-change, I will summeraize the other tools and link them

### Consensus Rule Change

If `block.number >= ISTANBUL_FORK_NUM` then when constructing the block, set a 16th value `blockHashesRoot` in the header to the `MMRRoot` from the previous block.

This Root is defined by the combining all previous hashes into a [Merkle Mountain Range](https://github.com/juinc/tilap/issues/244) (Peter Todd) using it's block number as leaf index.

Implementation of this tree is up to the clients but its fairly simple and requires only an additional ~500mb (on disk) to a full/archive node.

The MMR structure uses a proccess of "bagging the peaks" in order to create a single _root_ hash. However, this may be sub-optimal for our use case. I would like to ammend this to instead define each root as _the hash of the concatonation of the mountain peaks_. This will make building and verifying proofs much simpler without significantly changing the size of the proofs.

> The structure can be implimented as an array. Only a log(n) number of read/writes are done for any operation. A `put` operation needs to be done durring each block validation by the client. About 600 `get`s need to be performed when one of these proofs is requested (however the client can cache the latest proof). There is exactly 1 valid FlyProof for the current tip (latest mined block).

Reviewer 2: ```suggestion
The structure can be implemented as an array. Only a log(n) number of read/writes are done for any operation. A `put` operation needs to be done durring each block validation by the client. About 600 `get`s need to be performed when one of these proofs is requested (however the client can cache the latest proof). There is exactly 1 valid FlyProof for the current tip (latest mined block).
```

The FlyCLient paper can be followed almost exactly. Find it [here](https://eprint.iacr.org/2019/226.pdf)for specs.

Missing from the paper is a specific way to attain the samples from the blockhash (used as random seed). For this process, I propose we use [slice-sampling](https://en.wikipedia.org/wiki/Slice_sampling). This should be a very efficient algorithm as long as our probability density function is _invertible_. We may require a hash during each iteration (~600). This process is only required for building and verifying proofs, it is not required for instance, of miners.

<!--T

> A PoC JS implimentaion with the full, real dataset (7 million tree values) shows the following statistics:

Reviewer 2: ```suggestion
A PoC JS implementation with the full, real dataset (7 million tree values) shows the following statistics:
```

created: 2019-03-12
replaces (*optional): 210, 1094, 641

---

## Simple Summary

The addition of a Merkle-Mountain-Range (MMR) root to blockheaders

## Abstract
<!--A short (~200 word) description of the technical issue being addressed.-->

A block _chain_ could be interpreted as a linked-list using hashpointers from each block to its previous block. A Merkle-Mountain-Range is a type of tree that, will allow every block to point back to _all_ its previous blocks. This can be accomplished using only a single hashPointer in each block - the current MMR _root_. Adding this will allow proving the cummulative work of an entire chain in Olog(n) time, instead of the current O(n) time. In the case of ethereum, with small block-times, and over 7 million blocks and counting, this will open a ton of doors in the area light-clients, and data verification.

## Motivation
<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->

> The recent FlyClient [paper](https://eprint.iacr.org/2019/226.pdf) and [presentation](https://www.youtube.com/watch?v=BPNs9EVxWrA) at Scaling Bitcoin Represented a major breakthrough the ability to succinctly, and non-interactivly prove the cummulative work of a blockchain. 

Reviewer 3: Second link could be more precise: https://youtu.be/BPNs9EVxWrA?t=8390

Author: ```suggestion
The recent FlyClient [paper](https://eprint.iacr.org/2019/226.pdf) and [presentation](https://www.youtube.com/watch?v=BPNs9EVxWrA?t=8400) at Scaling Bitcoin Represented a major breakthrough the ability to succinctly, and non-interactivly prove the cumulative work of a blockchain.
```

<!-- at all unless you check its proof of work. Doing that would yeild a cost parameter:

- `block_x` is proven to have cost about `5 Ether` to produce.

nowing that something cost 5 Ether is not sufficient for most applications. By checking a suffix (blocks after this point in the chain) we can prove that the data cost at least `5n Ether` to produce, where n is the number of blocks checked.

With the simple assumption that the honest chain has 2/3 of the total mining power

The next thing we can do is to request the header before that and validate its PoW as well -->

The problem is that it ends there. And we still have nothing, if the person (or machine) validating cannot validate the block. The cost to produce such proofs is free; To validate blocks we cant use the same strategy. Blocks validate against a chain, which in turn validates through its proof of work. Currently the only way to do this requires 4gb of chain data. Thats way too much. We need something that can work within a few seconds. 

FlyClient is that something, and adding an `MMRroot` to each blockheader is the protocol change that will allow FlyClient to work.

> On a high level, it allows the verifier to sample a subset of blocks in a tree, rather than a full amount thought a linked list, improving the time-complexit from O(n) to O(log(n)). The sample subset can be made large enough such it is cryptographically infeasible to ever produce a false proof.

Reviewer 3: ```suggestion
On a high level, it allows the verifier to sample a subset of blocks in a tree, rather than a full amount thought a linked list, improving the time-complexity from O(n) to O(log(n)). The sample subset can be made large enough such it is cryptographically infeasible to ever produce a false proof.
```

## Specification
<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->

There is a consensus-breaking rule change proposed here, and then a bit of software around the change will be needed in order to take advantage of the new features. I think it would be best to keep _this_ EIP as just detailing the rule-change, I will summeraize the other tools and link them

### Consensus Rule Change

If `block.number >= ISTANBUL_FORK_NUM` then when constructing the block, set a 16th value `blockHashesRoot` in the header to the `MMRRoot` from the previous block.

This Root is defined by the combining all previous hashes into a [Merkle Mountain Range](https://github.com/juinc/tilap/issues/244) (Peter Todd) using it's block number as leaf index.

> Implimentaion of this tree is up to the clients but its fairly simple and requires only an additional ~500mb (on disk) to a full/archive node.

Reviewer 4: ```suggestion
Implementation of this tree is up to the clients but its fairly simple and requires only an additional ~500mb (on disk) to a full/archive node.
```

> The MMR structure uses a proccess of "bagging the peaks" in order to create a single _root_ hash. However, this may be sub-optimal for our use case. I would like to ammend this to instead define each root as _the hash of the concatonation of the mountain peaks_. This will make building and verifying proofs much simpler without significantly changing the size of the proofs.

Reviewer 3: ```suggestion
The MMR structure uses a process of "bagging the peaks" in order to create a single _root_ hash. However, this may be sub-optimal for our use case. I would like to amend this to instead define each root as _the hash of the concatenation of the mountain peaks_. This will make building and verifying proofs much simpler without significantly changing the size of the proofs.
```

It's not clear why "it may be sub-optimal for our use case". Is it about performance? Feasibility? Ease of use? Ease of implementation? Size?
In what aspect "hash of concatenation of the mountain peaks" is better than full tree structure (Todd's "bagged peaks" approach)? 
Is the "hash of concatenation" similar to the bagging method described in [grin docs](https://github.com/mimblewimble/grin/blob/master/doc/mmr.md)?

Author: this is a good question which i did not describe fully. @tzdybal, from [my js implementation](https://github.com/zmitton/merkle-mountain-range)

> The most elegant way of dealing with merkle-proofs is to basically think of them as a sparse tree in which only some of its nodes are present, and where a proof can be considered invalid iff, while traversing said tree, an undefined node is touched, or a node exists who's children do not hash to itself (more on this below). The second tree structure made from the peaks would not have the same property of being append only and theirfore, would not contain nodes that can be referenced by a perminent node index from the array. This means shared proofs would have to be less effiecient or more complex in some way.

so to what does a proof look like? it is the same data structure as the tree, except the "node database" only contains _some_ of the nodes. When i attempt to `get()` leaves from the tree, I will traverse the necessary nodes, which validate the proof as I go. 

For this to work properly, since i dont have all the nodes, I need to explicitly associate each node with its index position. The problem with "bagging the peaks" is that these "bag nodes" are different. They are not persistent, and do not fit the above abstraction. So in order to support such proofs I would have to pass more information (the  "bag nodes") with the proof.

Not only is this more complex, but when proving _multiple_ leaves at once, the proofs would be larger, because the bag node data can not be shared between leaves (unlike the regular database nodes). I will attempt to createa a picture below

Author: A tree with 12 leaves would look like this:
![image](https://user-images.githubusercontent.com/6954889/57118823-d53fd200-6d33-11e9-87fc-2fcbc1482e18.png)
If I wanted to verify the 5th leaf against a known root, I could use only the circled nodes. Also notice that at the moment leaf 5 was inserted, the peaks (needed to verify _its_ merkle root) were the blue circles (6 and 9), which nicely _happen_ to already be part of the later proof. if I also wanted to verify leaf 8, I would just add a few more nodes to the set: 15, 16, 17, and 20.

Using Todd's method, I would need more nodes (for instance a node that is the hash of 14 and 21) - "bag nodes". Those nodes are a separate set. This would complicate the relationship between leaf position, and the proof hashes.

Also, it would make multi-proofs less efficient. Since the peaks are constantly changing, the bag nodes would change too, and could not be completely shared between proofs the same way that the append-only ones are.

> The structure can be implimented as an array. Only a log(n) number of read/writes are done for any operation. A `put` operation needs to be done durring each block validation by the client. About 600 `get`s need to be performed when one of these proofs is requested (however the client can cash the latest proof). There is exactly 1 valid FlyProof for the current tip (latest mined block).

Reviewer 3: ```suggestion
The structure can be implimented as an array. Only a log(n) number of read/writes are done for any operation. A `put` operation needs to be done durring each block validation by the client. About 600 `get`s need to be performed when one of these proofs is requested (however the client can cache the latest proof). There is exactly 1 valid FlyProof for the current tip (latest mined block).
```
I would also consider using head instead of tip.

The FlyCLient paper can be followed almost exactly. Find it [here](https://eprint.iacr.org/2019/226.pdf)for specs.
 

> Missing from the paper is a specific way to attain the samples from the blockhash (used as random seed). For this proccess, I propose we use [slice-sampling](https://en.wikipedia.org/wiki/Slice_sampling). This should be a very efficient algorythm as long as our probability density function is _inversable_. We may require a hash during each iteration (~600). This proces is only required for building and verifying proofs, it is not required for instance, of miners.

Reviewer 3: ```suggestion
Missing from the paper is a specific way to attain the samples from the blockhash (used as random seed). For this process, I propose we use [slice-sampling](https://en.wikipedia.org/wiki/Slice_sampling). This should be a very efficient algorithm as long as our probability density function is _invertible_. We may require a hash during each iteration (~600). This process is only required for building and verifying proofs, it is not required for instance, of miners.
```
